src/object_accessor.hpp

```diff
@@ -268,75 +216,68 @@ Object Object::create(ContextType ctx, SharedRealm realm, const ObjectSchema &ob
 
     // populate
     Object object(realm, object_schema, table->get(row_index));
-    for (const Property& prop : object_schema.persisted_properties) {
+    if (out_row)
+        *out_row = object.row();
+    for (size_t i = 0; i < object_schema.persisted_properties.size(); ++i) {
+        auto& prop = object_schema.persisted_properties[i];
         if (prop.is_primary)
             continue;
 
-        if (Accessor::dict_has_value_for_key(ctx, value, prop.name)) {
-            object.set_property_value_impl(ctx, prop, Accessor::dict_value_for_key(ctx, value, prop.name), try_update);
+        auto v = ctx.value_for_property(value, prop.name, i);
+        if (!created && !v)
+            continue;
+
+        bool is_default = false;
+        if (!v) {
+            v = ctx.default_value_for_property(realm.get(), object_schema, prop.name);
+            is_default = true;
         }
-        else if (created) {
-            if (Accessor::has_default_value_for_property(ctx, realm.get(), object_schema, prop.name)) {
-                object.set_property_value_impl(ctx, prop, Accessor::default_value_for_property(ctx, realm.get(), object_schema, prop.name), try_update, true);
-            }
-            else if (prop.is_nullable || prop.type == PropertyType::Array) {
-                object.set_property_value_impl(ctx, prop, Accessor::null_value(ctx), try_update);
-            }
-            else {
+        if ((!v || ctx.is_null(*v)) && !prop.is_nullable && prop.type != PropertyType::Array) {
+            if (prop.is_primary || !ctx.allow_missing(value))
                 throw MissingPropertyValueException(object_schema.name, prop.name);
-            }
         }
+        if (v)
+            object.set_property_value_impl(ctx, prop, *v, try_update, is_default);
     }
     return object;
 }
 
 template<typename ValueType, typename ContextType>
-Object Object::get_for_primary_key(ContextType ctx, SharedRealm realm, const ObjectSchema &object_schema, ValueType primary_value)
+Object Object::get_for_primary_key(ContextType& ctx, SharedRealm realm,
```

Comment:
Since we're changing the signature of this method, we should consider changing these methods to take the Realm by (const) reference since there's no reason for `shared_ptr` to be involved here.